### ISSUE TITLE
Apparent inconsistency in request data encoding in multi_run with large number of reqs

### ISSUE THREAD

**Issue by gmbecker** [jeroen/curl#264]
It is possible I am just just misunderstanding how curl is supposed to work (or something else), I'm not an expert in libcurl, but the following "seems wrong" and I haven't been able to find an explanation for it. 

## Context:
I have a setup which chunks a (very) large request to an internal service into (potentially hundreds/low-thousands) small requests and then runs them all in very rapid succession. We get occasional (and difficult to to predict) failures which seem from the server side to be corrupted/malformed requests (where there is nothing wrong with the input data on the R side of things).

In attempting to debug that issue, I recently uncovered the following (which may or may not be the cause of the above), but either way seems to be wrong to my layman's understanding of curl:

## The Issue:

When creating a "large" number of handles and then asynchronously running them (via `multi_add` plus `multi_run`), the encoding on the (json) data of the request is not consistent. Sometimes it is plain text and sometimes it is json/plain text and sometimes it is `"data:application/octet-stream;base64"` There are no differences in the data being submitted that would explain this difference.

## Recreating the behavior:

For Politeness/good citizenship, this is done using the dockerized version http://httpbin.org:

` docker run -p 80:80 kennethreitz/httpbin`

Then do

```
library(jsonlite)
library(curl)
pool <- new_pool()

env <- new.env()

cb_fact <- function(env, nm) {
    force(env)
    force(nm)
    function(resp) {
        contchar <- rawToChar(resp$content)
        obj <- fromJSON(contchar)
        assign( nm, fromJSON(contchar), envir = env)
        TRUE
    }
}


make_handler <- function(nm) {
    json_query <- toJSON(list(mtcars = head(mtcars,2),  iris = head(iris, 2), name = nm))
    raw_query <- charToRaw(json_query)

    h <- new_handle(url = paste0("http://0.0.0.0:80/anything"),
                    customrequest="POST")
    handle_setheaders(h, "Content-Type" = "application/json", "Accept-Encoding" = "identity")
    handle_setopt(h, postfieldsize = length(raw_query), postfields = raw_query)
    h
}

for(i in 1:10) {
    for(j in 1:1000) {
        nm <- paste("result", i*1000 + j, sep = "_")
        h <- make_handler(nm)
        func <- cb_fact(env, nm)
        multi_add(h, done = func, pool = pool)
        if(length(multi_list(pool)) > 1000) {
            message("running partial set")
            multi_run(pool = pool)
        }
    }
}

multi_run(pool = pool)


lst <- as.list(env)

how_many_octet <- function(env)
    table(vapply(as.list(env), function(x) grepl("octet", x$data), NA))


```

On my local machine I got 

```
> how_many_octet(env)

FALSE  TRUE 
 4484  5516 
```

(the exact numbers vary *a lot*, but those numbers are real for one time I ran it; other times the TRUEs have gotten up near 85-90%).

Note that we are hitting the `/anything`  endpoint of httpbin, which returns exactly what it is given, and that we are specifying `"Accept-Encoding"=identity` in the handle headers, which means that, to my understanding, the data should be identical (and identically encoded) except for the "name" string I put in there. But it is not, with a bit over half being base64 stream encoded and a bit under half being plain-text encoding.

As a shot in the dark I tried turning multiplexing off and setting the number of concurrent host connections to 1,  but that didn't change the behavior from what I saw.

Also, note that if we do it the really really stupid inefficient way and do a `multi_run` after every single individual add, we do not see this behavior:

```

env2 <- new.env()
pool2 <- new_pool()
lapply(1:10000, function(i) {
    nm <- paste0("result_", i)
    if(i%%1000 == 0)
        message("On step ", i)
    h <- make_handler(nm)
    func <- cb_fact(env2, nm)
    multi_add(h, done = func, pool = pool2)
    multi_run(pool = pool2)
})
```

```
> how_many_octet(env2)

FALSE 
10000 
```

**Comment by jay** [comment]
Can you give an example of a query that is sent using the unexpected content type?

**Comment by gmbecker** [comment]
Hi @jay, I'm not entirely sure what you mean. The above (R) code hits lightweight wrappers around libcurl, so my understanding is the concept of handles, pools, etc comes straight form there.

It also does not seem to matter too much what the data of the request is (though it is JSON in both the internal application and the baby example here). 

In the real application, the json data is two vectors (ordered sets, though in practice the order doesn't matter) which would be of the form, e.g., 
```
{"name1":["hi","there","you"],"name2":["what","is","up"],"thing3":"stuff","thing4":"things"} 
```
though name1 and name2 values would typically be longer.

**Comment by jay** [comment]
I mean an actual unexpected/malformed request that libcurl sends to the server. Are you able to capture the request with verbose mode or debug mode? I don't speak R. I am a developer on the libcurl project and trying to understand what you reported.

**Comment by gmbecker** [comment]
@jay, I see. Thanks for responding. I don't own the server application but I believe we did get logging on there turned on, and I will see if i can get a request out of that tomorrow.

I can't share the code (client or server) for the real application, but the above is closely adapted from it on the server side.

In case it helps, here is what httpbin gives back in the octet and non-octet versions (technically its R output  but parsed directly from the curl response so maybe it is helpful:

non-octet:
```
$args
named list()

$data
[1] "{\"mtcars\":[{\"mpg\":21,\"cyl\":6,\"disp\":160,\"hp\":110,\"drat\":3.9,\"wt\":2.62,\"qsec\":16.46,\"vs\":0,\"am\":1,\"gear\":4,\"carb\":4,\"_row\":\"Mazda RX4\"},{\"mpg\":21,\"cyl\":6,\"disp\":160,\"hp\":110,\"drat\":3.9,\"wt\":2.875,\"qsec\":17.02,\"vs\":0,\"am\":1,\"gear\":4,\"carb\":4,\"_row\":\"Mazda RX4 Wag\"}],\"iris\":[{\"Sepal.Length\":5.1,\"Sepal.Width\":3.5,\"Petal.Length\":1.4,\"Petal.Width\":0.2,\"Species\":\"setosa\"},{\"Sepal.Length\":4.9,\"Sepal.Width\":3,\"Petal.Length\":1.4,\"Petal.Width\":0.2,\"Species\":\"setosa\"}],\"name\":[\"result_96404\"]}"

$files
named list()

$form
named list()

$headers
$headers$Accept
[1] "*/*"

$headers$`Accept-Encoding`
[1] "identity"

$headers$`Content-Length`
[1] "486"

$headers$`Content-Type`
[1] "application/json"

$headers$Host
[1] "0.0.0.0"

$headers$`User-Agent`
[1] "R (4.1.2 x86_64-apple-darwin17.0 x86_64 darwin17.0)"


$json
$json$iris
  Petal.Length Petal.Width Sepal.Length Sepal.Width Species
1          1.4         0.2          5.1         3.5  setosa
2          1.4         0.2          4.9         3.0  setosa

$json$mtcars
              am carb cyl disp drat gear  hp mpg  qsec vs    wt
Mazda RX4      1    4   6  160  3.9    4 110  21 16.46  0 2.620
Mazda RX4 Wag  1    4   6  160  3.9    4 110  21 17.02  0 2.875

$json$name
[1] "result_96404"


$method
[1] "POST"

$origin
[1] "172.17.0.1"

$url
[1] "http://0.0.0.0/anything"

```

datastream octet:

```
$args
named list()

$data
[1] "data:application/octet-stream;base64,AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAUOJRf/d/AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIZeSd938AAOBSgL/3fwAA4FKAv/d/AAACAAAAAQAAAAAAAAAAAAAAQ29udGVudC1UeXBlOiBhcHBsaWNhdGlvbi9qc29uAABBY2NlcHQtRW5jb2Rpbmc6IGlkZW50aXR5AAAAAAAAACCkR5/3fwAAYKRHn/d/AABFeHBlY3Q6IAAAAAAAAADQUKRHn/d/AAAAAAAAAAAAAGRlZmxhdGUsIGd6aXAAANAvZXRjL3NzbC9jZXJ0LnBlbQAAAAA="

$files
named list()

$form
named list()

$headers
$headers$Accept
[1] "*/*"

$headers$`Accept-Encoding`
[1] "identity"

$headers$`Content-Length`
[1] "485"

$headers$`Content-Type`
[1] "application/json"

$headers$Host
[1] "0.0.0.0"

$headers$`User-Agent`
[1] "R (4.1.2 x86_64-apple-darwin17.0 x86_64 darwin17.0)"


$json
NULL

$method
[1] "POST"

$origin
[1] "172.17.0.1"

$url
[1] "http://0.0.0.0/anything"
```

I will try to get more debugging info tomorrow

**Comment by jay** [comment]
Please don't share anything sensitive. I meant the raw data from your repro with verbose = TRUE or something.

> [1] "data:application/octet-stream;base64,AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAUOJRf/d/AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIZeSd938AAOBSgL/3fwAA4FKAv/d/AAACAAAAAQAAAAAAAAAAAAAAQ29udGVudC1UeXBlOiBhcHBsaWNhdGlvbi9qc29uAABBY2NlcHQtRW5jb2Rpbmc6IGlkZW50aXR5AAAAAAAAACCkR5/3fwAAYKRHn/d/AABFeHBlY3Q6IAAAAAAAAADQUKRHn/d/AAAAAAAAAAAAAGRlZmxhdGUsIGd6aXAAANAvZXRjL3NzbC9jZXJ0LnBlbQAAAAA="

It looks like corrupt data was base64 encoded. I still can't tell if this is an rcurl issue or libcurl issue.

~~~
������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������PâQ÷��������������������������eä÷��àR€¿÷��àR€¿÷����������������Content-Type: application/json��Accept-Encoding: identity������� ¤GŸ÷��`¤GŸ÷��Expect: �������ÐP¤GŸ÷����������deflate, gzip��Ð/etc/ssl/cert.pem����
~~~

**Comment by gmbecker** [comment]
Hi @jay 

I had a chance to rerun things with verbose on and I can't say I understand much of any of this but here  is (the beginning of) what comes out. Apologies for the verbosity but as I'm not sure what you're looking for I didn't want to remove things, even things that seem to be repetitions with no new information.

The following is for 10 rapid fire requests, of which 3 seem to have gone bad. I don't know of any way to tell which ones the bad ones were from the output, unfortunately, but I hope this is something that helps. I can tell you  That the 4th, 5th, and 6th requests were the ones that were bad _**in the order they were added to the request pool**_ but I don't know if that order is reflected in the output.

```
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [serially]
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [serially]
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [serially]
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [serially]
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [serially]
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [serially]
* No more connections allowed to host: 6
* No connections available.
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [serially]
* No more connections allowed to host: 6
* No connections available.
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [serially]
* No more connections allowed to host: 6
* No connections available.
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [serially]
* No more connections allowed to host: 6
* No connections available.
*   Trying 0.0.0.0...
* TCP_NODELAY set
* Hostname '0.0.0.0' was found in DNS cache
*   Trying 0.0.0.0...
* TCP_NODELAY set
* Hostname '0.0.0.0' was found in DNS cache
*   Trying 0.0.0.0...
* TCP_NODELAY set
* Hostname '0.0.0.0' was found in DNS cache
*   Trying 0.0.0.0...
* TCP_NODELAY set
* Hostname '0.0.0.0' was found in DNS cache
*   Trying 0.0.0.0...
* TCP_NODELAY set
* Hostname '0.0.0.0' was found in DNS cache
*   Trying 0.0.0.0...
* TCP_NODELAY set
* Connected to 0.0.0.0 (127.0.0.1) port 80 (#0)
> POST /anything HTTP/1.1
Host: 0.0.0.0
User-Agent: R (4.1.2 x86_64-apple-darwin17.0 x86_64 darwin17.0)
Accept: */*
Content-Type: application/json
Accept-Encoding: identity
Content-Length: 483

* upload completely sent off: 483 out of 483 bytes
* Connected to 0.0.0.0 (127.0.0.1) port 80 (#1)
> POST /anything HTTP/1.1
Host: 0.0.0.0
User-Agent: R (4.1.2 x86_64-apple-darwin17.0 x86_64 darwin17.0)
Accept: */*
Content-Type: application/json
Accept-Encoding: identity
Content-Length: 483

* upload completely sent off: 483 out of 483 bytes
* Connected to 0.0.0.0 (127.0.0.1) port 80 (#2)
> POST /anything HTTP/1.1
Host: 0.0.0.0
User-Agent: R (4.1.2 x86_64-apple-darwin17.0 x86_64 darwin17.0)
Accept: */*
Content-Type: application/json
Accept-Encoding: identity
Content-Length: 483

* upload completely sent off: 483 out of 483 bytes
< HTTP/1.1 200 OK
< Server: gunicorn/19.9.0
< Date: Wed, 02 Mar 2022 00:19:20 GMT
< Connection: keep-alive
< Content-Type: application/json
< Content-Length: 3309
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
* Connection #2 to host 0.0.0.0 left intact
* Connected to 0.0.0.0 (127.0.0.1) port 80 (#3)
> POST /anything HTTP/1.1
Host: 0.0.0.0
User-Agent: R (4.1.2 x86_64-apple-darwin17.0 x86_64 darwin17.0)
Accept: */*
Content-Type: application/json
Accept-Encoding: identity
Content-Length: 483

* upload completely sent off: 483 out of 483 bytes
* Connected to 0.0.0.0 (127.0.0.1) port 80 (#4)
> POST /anything HTTP/1.1
Host: 0.0.0.0
User-Agent: R (4.1.2 x86_64-apple-darwin17.0 x86_64 darwin17.0)
Accept: */*
Content-Type: application/json
Accept-Encoding: identity
Content-Length: 483

* upload completely sent off: 483 out of 483 bytes
* Connected to 0.0.0.0 (127.0.0.1) port 80 (#5)
> POST /anything HTTP/1.1
Host: 0.0.0.0
User-Agent: R (4.1.2 x86_64-apple-darwin17.0 x86_64 darwin17.0)
Accept: */*
Content-Type: application/json
Accept-Encoding: identity
Content-Length: 483

* upload completely sent off: 483 out of 483 bytes
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [can pipeline]
* Re-using existing connection! (#2) with host 0.0.0.0
* Connected to 0.0.0.0 (127.0.0.1) port 80 (#2)
> POST /anything HTTP/1.1
Host: 0.0.0.0
User-Agent: R (4.1.2 x86_64-apple-darwin17.0 x86_64 darwin17.0)
Accept: */*
Content-Type: application/json
Accept-Encoding: identity
Content-Length: 483

* upload completely sent off: 483 out of 483 bytes
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [can pipeline]
* No more connections allowed to host: 6
* No connections available.
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [can pipeline]
* No more connections allowed to host: 6
* No connections available.
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [can pipeline]
* No more connections allowed to host: 6
* No connections available.
< HTTP/1.1 200 OK
< Server: gunicorn/19.9.0
< Date: Wed, 02 Mar 2022 00:19:20 GMT
< Connection: keep-alive
< Content-Type: application/json
< Content-Length: 1930
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
* Connection #0 to host 0.0.0.0 left intact
< HTTP/1.1 200 OK
< Server: gunicorn/19.9.0
< Date: Wed, 02 Mar 2022 00:19:20 GMT
< Connection: keep-alive
< Content-Type: application/json
< Content-Length: 1930
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
* Connection #1 to host 0.0.0.0 left intact
< HTTP/1.1 200 OK
< Server: gunicorn/19.9.0
< Date: Wed, 02 Mar 2022 00:19:20 GMT
< Connection: keep-alive
< Content-Type: application/json
< Content-Length: 1092
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
* Connection #3 to host 0.0.0.0 left intact
< HTTP/1.1 200 OK
< Server: gunicorn/19.9.0
< Date: Wed, 02 Mar 2022 00:19:20 GMT
< Connection: keep-alive
< Content-Type: application/json
< Content-Length: 1092
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
* Connection #4 to host 0.0.0.0 left intact
< HTTP/1.1 200 OK
< Server: gunicorn/19.9.0
< Date: Wed, 02 Mar 2022 00:19:20 GMT
< Connection: keep-alive
< Content-Type: application/json
< Content-Length: 1092
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
* Connection #5 to host 0.0.0.0 left intact
< HTTP/1.1 200 OK
< Server: gunicorn/19.9.0
< Date: Wed, 02 Mar 2022 00:19:20 GMT
< Connection: keep-alive
< Content-Type: application/json
< Content-Length: 1930
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
* Connection #2 to host 0.0.0.0 left intact
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [can pipeline]
* Re-using existing connection! (#5) with host 0.0.0.0
* Connected to 0.0.0.0 (127.0.0.1) port 80 (#5)
> POST /anything HTTP/1.1
Host: 0.0.0.0
User-Agent: R (4.1.2 x86_64-apple-darwin17.0 x86_64 darwin17.0)
Accept: */*
Content-Type: application/json
Accept-Encoding: identity
Content-Length: 483

* upload completely sent off: 483 out of 483 bytes
< HTTP/1.1 200 OK
< Server: gunicorn/19.9.0
< Date: Wed, 02 Mar 2022 00:19:20 GMT
< Connection: keep-alive
< Content-Type: application/json
< Content-Length: 1930
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
* Connection #5 to host 0.0.0.0 left intact
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [can pipeline]
* Re-using existing connection! (#5) with host 0.0.0.0
* Connected to 0.0.0.0 (127.0.0.1) port 80 (#5)
> POST /anything HTTP/1.1
Host: 0.0.0.0
User-Agent: R (4.1.2 x86_64-apple-darwin17.0 x86_64 darwin17.0)
Accept: */*
Content-Type: application/json
Accept-Encoding: identity
Content-Length: 483

* upload completely sent off: 483 out of 483 bytes
* Found bundle for host 0.0.0.0: 0x7fb0cdc09b90 [can pipeline]
* Re-using existing connection! (#4) with host 0.0.0.0
* Connected to 0.0.0.0 (127.0.0.1) port 80 (#4)
> POST /anything HTTP/1.1
Host: 0.0.0.0
User-Agent: R (4.1.2 x86_64-apple-darwin17.0 x86_64 darwin17.0)
Accept: */*
Content-Type: application/json
Accept-Encoding: identity
Content-Length: 483

* upload completely sent off: 483 out of 483 bytes
< HTTP/1.1 200 OK
< Server: gunicorn/19.9.0
< Date: Wed, 02 Mar 2022 00:19:20 GMT
< Connection: keep-alive
< Content-Type: application/json
< Content-Length: 1930
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
* Connection #5 to host 0.0.0.0 left intact
< HTTP/1.1 200 OK
< Server: gunicorn/19.9.0
< Date: Wed, 02 Mar 2022 00:19:20 GMT
< Connection: keep-alive
< Content-Type: application/json
< Content-Length: 1930
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
* Connection #4 to host 0.0.0.0 left intact
```

**Comment by jay** [comment]
> The following is for 10 rapid fire requests, of which 3 seem to have gone bad.

I see 10 POST requests with the same headers and same content length. I see 10 replies 200 OK with the same headers but different timestamps (expected) and different content lengths.
~~~
< Content-Length: 3309
< Content-Length: 1930
< Content-Length: 1930
< Content-Length: 1092
< Content-Length: 1092
< Content-Length: 1092
< Content-Length: 1930
< Content-Length: 1930
< Content-Length: 1930
< Content-Length: 1930
~~~

I would expect the content lengths could be a byte or two different since the json includes the counter. I assume some or all of those replies are corrupt data. Is it possible that the postfields data raw_query is being destroyed or garbage collected after makehandler returns? (As I've mentioned I don't know about R and how it works) What happens if you use copypostfields ([CURLOPT_COPYPOSTFIELDS](https://curl.se/libcurl/c/CURLOPT_COPYPOSTFIELDS.html)) instead of postfields ([CURLOPT_POSTFIELDS](https://curl.se/libcurl/c/CURLOPT_POSTFIELDS.html))? Note the size must be set before the copy.

~~~R
handle_setopt(h, postfieldsize = length(raw_query), copypostfields = raw_query)
~~~

I don't know if R handle_setopt guarantees postfieldsize is set before copypostfields, I don't see it in the doc, so if that doesn't work try

~~~R
handle_setopt(h, postfieldsize = length(raw_query))
# postfieldsize must be set before calling copypostfields to copy the data
handle_setopt(h, copypostfields = raw_query)
~~~

and if that solves it in your stress testing, it would be more efficient if you have large amounts of data to not copy and instead find a way to keep raw_query from being garbage collected... tie it to h or something?

**Comment by gmbecker** [comment]
@jay thank you for your help. Yes, changing postfields to copypostfields does indeed fix the issue.

I had no idea doing `handle_setopt(h, postfieldsize = length(raw_query), postfields = raw_query)` would try to lazily consume postfield, and I suspect neither did R, so the object was not protected from garbage collection.

This seems like a bug in the R curl package: when passed an R object as postfields, it should be preserving the object for the life of the handle, and it seems to not be doing so.

**Comment by jeroen** [comment]
Thanks that is a good catch! I think the safest solution is to always use copypostfields instead of postfields, so that we don't have to worry if the GC is going to eat the input before curl can read it.

**Comment by jay** [comment]
> I think the safest solution is to always use copypostfields instead of postfields, so that we don't have to worry if the GC is going to eat the input before curl can read it.

It appears your code always sets postfieldsize before setting postfields (now copypostfields), however then I wonder what is the point of postfieldsize if it's always overridden internally.

I would have expected that postfieldsize is not set unless explicitly done by the user, but if that was the case and postfields is actually copypostfields then it may cause unexpected behavior if the user had set postfieldsize after setting postfields.

In other words , in libcurl this will do what the user expected:

~~~c
curl_easy_setopt(curl, CURLOPT_POSTFIELDS, mem);
curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE_LARGE, (curl_off_t)memsize);
~~~

~~~c
curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE_LARGE, (curl_off_t)memsize);
curl_easy_setopt(curl, CURLOPT_POSTFIELDS, mem);
~~~

~~~c
curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE_LARGE, (curl_off_t)memsize);
curl_easy_setopt(curl, CURLOPT_COPYPOSTFIELDS, mem);
~~~

but this usually will not:

~~~c
// this copies mem and since the size is not set libcurl internally will set the post size to strlen(mem) and then copy that amount
curl_easy_setopt(curl, CURLOPT_COPYPOSTFIELDS, mem);
// this sets the post size, but the size was already set so this actually overrides the size
// while technically this is valid (eg copy large amount but post smaller amount) it is usually not what the user intended
curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE_LARGE, (curl_off_t)memsize);
~~~

**Comment by jay** [comment]
> however then I wonder what is the point of postfieldsize if it's always overridden internally.

nevermind it looks like allowing it to be set later just allows for scenario 4 in my previous message, essentially you give the user the option to send less data.

**Comment by jeroen** [comment]
Yes there are several option that the user should not set manually, because they are implicit in some other option. Another example are fields like `CURLOPT_XFERINFODATA`, `CURLOPT_READDATA`, etc, which are implicit when the user is setting `CURLOPT_XFERINFOFUNCTION` or `CURLOPT_READFUNCTION`

This has never caused problems but perhaps it would be safer to explicitly ignore those options.

**Comment by jay** [comment]
> This has never caused problems but perhaps it would be safer to explicitly ignore those options.

The thing about scenario 4 as I explained it is I did not account for a binding so it's sort of misleading. If you in your binding set the size to the total size of the data, the data is copied and then the user later sets it to the size they want to send, that is fine and what the user expected. In my opinion the unexpected case for R curl is then if the user set the size they wanted to send and actually some larger size (the initial size set in the binding) is sent.

Note also libcurl has a safety check setting the size when CURLOPT_COPYPOSTFIELDS was called prior that will unset the data if the user specifies a size larger than the size copied.

https://github.com/curl/curl/blob/curl-7_81_0/lib/setopt.c#L522-L539
